### PR TITLE
add a more advanced Finder configuration support

### PR DESCRIPTION
**Pull request by sstok**
This a Proof of concept to introduce a more advanced Configuration for the Finder.

Basically to use the finder you use the following configuration:

``` yaml
finder:
    in: [src, tests]
    exclude:
        - fixtures
        - stubs
    not_name:
        - ".blade.php"
    not_contains:
        - "Kernel"
```

Tests are missing and PHPDoc is incomplete, as I would to have some early feedback first ;)

Related to https://github.com/StyleCI/StyleCI/issues/270

### CODE REVIEW

File: src/ConfigFactory.php [hunk 1]
```diff
@@ -47,14 +61,44 @@ public function make(array $input = [])
...
     }
 
     /**
+     * Make a new finder config object.
+     *
+     * @param array $input
+     *
+     * @throws \StyleCI\Config\Exceptions\InvalidFinderType
+     *
+     * @return FinderConfig
```
Comment (GrahamCampbell) [inline src/ConfigFactory.php]: please use the FQCN

File: src/ConfigFactory.php [hunk 2]
```diff
@@ -22,6 +23,19 @@
  */
```
Comment (GrahamCampbell) [inline src/ConfigFactory.php]: Feel free to add yourself to the authors list of this class. :)

**Comment by GrahamCampbell** [comment]
Looks quite nice. :)

**Comment by sstok** [comment]
Ok, I will continue with the pr then and fix the mentioned issues ;)

**Comment by GrahamCampbell** [comment]
Awesome. :)

**Comment by joecohens** [comment]
:+1:

**Comment by sstok** [comment]
Done, ready for review.

File: src/ConfigFactory.php (hunk 1 above)
Comment (joecohens) [inline src/ConfigFactory.php]: @sstok `@return \StyleCI\Config\FinderConfig`

Comment (sstok) [inline src/ConfigFactory.php]: Done

File: src/Config.php [hunk 3]
```diff
@@ -384,6 +385,13 @@ class Config
     protected $excluded = [];
 
     /**
+     * The configuration of the Finder.
+     *
+     * @var FinderConfig
```
Comment (GrahamCampbell) [inline src/Config.php]: please use the FQCN in the docblocks for human readability :)

Comment (GrahamCampbell) [inline src/Config.php]: also, should be `|null` too :)

**Comment by GrahamCampbell** [comment]
The FinderConfig class probably needs more test coverage, though, this looks pretty good, thank you!.

Pinging @joecohens, @jbrooksuk for review.

**Comment by sstok** [comment]
Renamed InvalidFinderDirectory to InvalidFinderDirectoryException, same for InvalidFinderType.
Used a more strict testing for "in", InvalidFinderDirectoryException now accepts a none-string value (same as the other exceptions).

File: src/Config.php (hunk 3 above)
Comment (sstok) [inline src/Config.php]: Done :+1:

File: src/FinderConfig.php [hunk 4]
```diff
@@ -0,0 +1,397 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace StyleCI\Config;
+
+use StyleCI\Config\Exceptions\InvalidFinderDirectoryException;
+
+/**
+ * This is the Finder configuration class.
```
Comment (GrahamCampbell) [inline src/FinderConfig.php]: lower case f please

File: src/FinderConfig.php [hunk 5]
```diff
@@ -0,0 +1,397 @@
...
+    {
+        $this->exclude = (array) $dirs;
+
+        return $this;
+    }
+
+    /**
+     * Set the rules that files must match.
+     *
+     * @param array|string $patters
```
Comment (GrahamCampbell) [inline src/FinderConfig.php]: should we not have `string[]` rather than `array`?

Comment (jbrooksuk) [inline src/FinderConfig.php]: Yeah.

**Comment by jbrooksuk** [comment]
Looks really good to me! Seems to cover all use cases.

File: src/FinderConfig.php (hunk 5 above)
Comment (sstok) [inline src/FinderConfig.php]: Done.

Comment (GrahamCampbell) [inline src/FinderConfig.php]: Thanks.

**Comment by GrahamCampbell** [comment]
Just to check, this isn't breaking with the current version?

**Comment by sstok** [comment]
It should be compatible with the current version, only when you configure the finder the "exclude" and "extensions" are ignored.

**Comment by joecohens** [comment]
Looks very nice!

**Comment by GrahamCampbell** [comment]
Thank you so much!

**Comment by sstok** [comment]
No problem ;) I will update the Fixer tomorrow to use the new FinderConfig :+1:

**Comment by sstok** [comment]
https://github.com/StyleCI/Fixer/pull/7
